File: sample/flutterApp/lib/src/support/sample_event_card.dart
// Copyright © 2026 Shiva Thapa (@shivathapaa). All rights reserved.
// Use of this source code is governed by the MPL 2.0 license that can be
// found in the LICENSE file.

import 'package:flutter/material.dart';
import 'package:nepali_date_picker_kmp/nepali_date_picker_kmp.dart';

import 'sample_event_payload.dart';

/// A tapped event as a product would draw it, built from the JSON the event
/// carries in its payload.
///
/// The bridge hands the event back exactly as it was given, id and payload
/// untouched, and everything below the name here is this sample reading its own
/// record out of that string. The image links are printed rather than fetched:
/// loading remote images needs an image package, and the point here is the
/// payload, not the network.
class SampleEventCard extends StatelessWidget {
  const SampleEventCard({super.key, required this.event});

  final NepaliEvent event;

  @override
  Widget build(BuildContext context) {
    final payload = event.samplePayload;
    final accent = payload?.accent ?? Theme.of(context).colorScheme.primary;
    final labels = Theme.of(context).textTheme.labelSmall;

    return Card(
      clipBehavior: Clip.antiAlias,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Container(
            width: double.infinity,
            color: accent.withValues(alpha: 0.25),
            padding: const EdgeInsets.all(12),
            child: Row(
              children: [
                if (payload != null)
                  Padding(
                    padding: const EdgeInsets.only(right: 8),
                    child: Icon(payload.iconData, color: accent),
                  ),
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(event.name,
                          style: Theme.of(context).textTheme.titleMedium),
                      Text(
                        '${event.kind.name} · ${event.year}/${event.month}/${event.dayOfMonth}',
                        style: labels,
                      ),
                    ],
                  ),
                ),
              ],
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(12),
            child: payload == null
                ? const Text('This event carries no payload.')
                : Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    spacing: 6,
                    children: [
                      Text(payload.description),
                      if (payload.venue != null)
                        _line('Where', payload.venue!, labels),
                      if (payload.organizer != null)
                        _line('Who', payload.organizer!, labels),
                      if (payload.bannerUrl != null)
                        _line('Banner', payload.bannerUrl!, labels),
                      if (payload.iconUrl != null)
                        _line('Icon', payload.iconUrl!, labels),
                      if (payload.link != null)
                        _line('Link', payload.link!, labels),
                      _line('Correlates to', event.id ?? 'no id', labels),
                      if (payload.tags.isNotEmpty)
                        Wrap(
                          spacing: 6,
                          children: [
                            for (final tag in payload.tags)
                              Chip(
                                label: Text(tag),
                                backgroundColor: accent.withValues(alpha: 0.2),
                                visualDensity: VisualDensity.compact,
                              ),
                          ],
                        ),
                    ],
                  ),
          ),
        ],
      ),
    );
  }

  Widget _line(String label, String value, TextStyle? style) => Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text('$label ',
              style: style?.copyWith(fontWeight: FontWeight.w600)),
          Expanded(
            child: Text(value,
                style: style, maxLines: 1, overflow: TextOverflow.ellipsis),
          ),
        ],
      );
}

File: sample/composeApp/src/commonMain/kotlin/dev/shivathapaa/nepalidatepicker/SampleEventCard.kt
/*
 * Copyright © 2026 Shiva Thapa (@shivathapaa). All rights reserved.
 *
 * Licensed under the Mozilla Public License, Version 2.0 (the "License").
 */

package dev.shivathapaa.nepalidatepicker

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import dev.shivathapaa.nepalidatepickerkmp.event.NepaliCalendarEvent

private val CardCornerRadius = 12.dp
private val CardPadding = 12.dp
private val BannerHeight = 56.dp
private val IconSize = 24.dp
private val RowGap = 6.dp
private val ChipGap = 6.dp
private val ChipCornerRadius = 8.dp
private val ChipPadding = 6.dp
private const val BannerAlpha = 0.25f

/**
 * A tapped event as a product would draw it, built from the JSON the event carries in its payload.
 *
 * The library hands back the event it was given, id and payload untouched, and everything below the
 * name here, the colour, the icon, the description, the tags, the venue and the links, is this
 * sample reading its own record out of that string.
 *
 * The image links are printed rather than fetched: loading remote images needs an image library,
 * and the point here is the payload, not the network.
 */
@Composable
fun SampleEventCard(event: NepaliCalendarEvent, modifier: Modifier = Modifier) {
    val payload = event.samplePayload()
    val accent = payload?.accentColor() ?: MaterialTheme.colorScheme.primary

    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(CardCornerRadius),
        color = MaterialTheme.colorScheme.surfaceVariant
    ) {
        Column {
            Box(
                modifier = Modifier.fillMaxWidth()
                    .height(BannerHeight)
                    .background(accent.copy(alpha = BannerAlpha))
            ) {
                Row(
                    modifier = Modifier.align(Alignment.CenterStart).padding(horizontal = CardPadding),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(RowGap)
                ) {
                    if (payload != null) {
                        Icon(
                            imageVector = payload.iconVector(),
                            contentDescription = null,
                            tint = accent,
                            modifier = Modifier.size(IconSize)
                        )
                    }
                    Column {
                        Text(
                            text = event.name,
                            style = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.SemiBold
                        )
                        Text(
                            text = "${event.kind.name} · ${event.date.formatted()}",
                            style = MaterialTheme.typography.labelSmall
                        )
                    }
                }
            }

            Column(
                modifier = Modifier.padding(CardPadding),
                verticalArrangement = Arrangement.spacedBy(RowGap)
            ) {
                if (payload == null) {
                    Text(
                        text = "This event carries no payload, so there is nothing to read out of it.",
                        style = MaterialTheme.typography.bodySmall
                    )
                    return@Column
                }

                Text(text = payload.description, style = MaterialTheme.typography.bodyMedium)
                payload.venue?.let { PayloadLine(label = "Where", value = it) }
                payload.organizer?.let { PayloadLine(label = "Who", value = it) }
                payload.bannerUrl?.let { PayloadLine(label = "Banner", value = it) }
                payload.iconUrl?.let { PayloadLine(label = "Icon", value = it) }
                payload.link?.let { PayloadLine(label = "Link", value = it) }
                PayloadLine(label = "Correlates to", value = event.id ?: "no id")

                if (payload.tags.isNotEmpty()) {
                    FlowRow(horizontalArrangement = Arrangement.spacedBy(ChipGap)) {
                        payload.tags.forEach { tag -> PayloadChip(tag = tag, accent = accent) }
                    }
                }
            }
        }
    }
}

/** One labelled line of the record, kept to a single line so a long URL does not push the card. */
@Composable
private fun PayloadLine(label: String, value: String) {
    Row(horizontalArrangement = Arrangement.spacedBy(RowGap)) {
        Text(
            text = label,
            style = MaterialTheme.typography.labelSmall,
            fontWeight = FontWeight.Medium
        )
        Text(
            text = value,
            style = MaterialTheme.typography.labelSmall,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis
        )
    }
}

/** One of the payload's own tags, drawn in the colour the payload asked for. */
@Composable
private fun PayloadChip(tag: String, accent: Color) {
    Surface(
        shape = RoundedCornerShape(ChipCornerRadius),
        color = accent.copy(alpha = BannerAlpha)
    ) {
        Text(
            text = tag,
            modifier = Modifier.padding(horizontal = ChipPadding, vertical = ChipGap / 2),
            style = MaterialTheme.typography.labelSmall
        )
    }
}
